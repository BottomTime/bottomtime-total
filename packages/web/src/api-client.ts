import { ApiClient } from '@bottomtime/api';

import { InjectionKey, inject } from 'vue';

export const ApiClientKey: InjectionKey<ApiClient> = Symbol('ApiClient');
export function useClient(): ApiClient {
  const client = inject(ApiClientKey);

  if (!client) {
    throw new Error(
      'Api Client has not been initialized. Did you forget to call `app.provide(key, client)`?',
    );
  }

  return client;
}
