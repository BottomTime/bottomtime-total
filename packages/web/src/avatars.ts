import { AvatarSize, ListAvatarURLsResponseDTO } from '@bottomtime/api';

export type AvatarURLs = ListAvatarURLsResponseDTO['sizes'];

export function getAvatarURL(
  baseUrl: string | undefined,
  size: AvatarSize,
): string | undefined {
  if (!baseUrl) return undefined;

  if (/^\/api\/users\/[a-z0-9]+([_.-][a-z0-9]+)*\/avatar/i.test(baseUrl)) {
    return baseUrl.endsWith('/') ? `${baseUrl}${size}` : `${baseUrl}/${size}`;
  }

  return baseUrl;
}
