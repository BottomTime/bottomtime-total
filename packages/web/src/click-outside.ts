import { Directive, DirectiveBinding } from 'vue';

type HTMLElementWithClickOutside = HTMLElement & {
  clickOutsideEvent: (e: MouseEvent) => void;
};

export const clickOutside: Directive = {
  beforeMount: (el: HTMLElementWithClickOutside, binding: DirectiveBinding) => {
    el.clickOutsideEvent = (event) => {
      // here I check that click was outside the el and his children
      if (!(el == event.target || el.contains(event.target as Node))) {
        // and if it did, call method provided in attribute value
        binding.value();
      }
    };
    document.addEventListener('click', el.clickOutsideEvent);
  },
  unmounted: (el: HTMLElementWithClickOutside) => {
    document.removeEventListener('click', el.clickOutsideEvent);
  },
};
