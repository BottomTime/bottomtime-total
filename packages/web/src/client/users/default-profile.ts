import { assertValid } from '@/helpers';
import { SuperAgentStatic } from 'superagent';
import {
  Profile,
  ProfileCertificationData,
  ProfileData,
  ProfileDataSchema,
  UserData,
} from './interfaces';
import { ProfileVisibility } from '@/constants';

export class DefaultProfile implements Profile {
  private static readonly JsonOmitKeys = new Set([
    'userId',
    'memberSince',
    'username',
  ]);

  constructor(
    private readonly agent: SuperAgentStatic,
    private readonly data: UserData,
  ) {}

  get userId(): string {
    return this.data.id;
  }
  get memberSince(): Date {
    return this.data.memberSince;
  }
  get username(): string {
    return this.data.username;
  }

  get avatar(): string | undefined {
    return this.data.profile.avatar;
  }
  set avatar(value: string | undefined) {
    this.data.profile.avatar = value;
  }

  get bio(): string | undefined {
    return this.data.profile.bio;
  }
  set bio(value: string | undefined) {
    this.data.profile.bio = value;
  }

  get birthdate(): string | undefined {
    return this.data.profile.birthdate;
  }
  set birthdate(value: string | undefined) {
    this.data.profile.birthdate = value;
  }

  get customData(): Record<string, unknown> | undefined {
    return this.data.profile.customData;
  }
  set customData(value: Record<string, unknown> | undefined) {
    this.data.profile.customData = value;
  }

  get certifications(): ProfileCertificationData[] | undefined {
    return this.data.profile.certifications;
  }
  set certifications(value: ProfileCertificationData[] | undefined) {
    this.data.profile.certifications = value;
  }

  get experienceLevel(): string | undefined {
    return this.data.profile.experienceLevel;
  }
  set experienceLevel(value: string | undefined) {
    this.data.profile.experienceLevel = value;
  }

  get location(): string | undefined {
    return this.data.profile.location;
  }
  set location(value: string | undefined) {
    this.data.profile.location = value;
  }

  get name(): string | undefined {
    return this.data.profile.name;
  }
  set name(value: string | undefined) {
    this.data.profile.name = value;
  }

  get profileVisibility(): ProfileVisibility {
    return this.data.profile.profileVisibility;
  }
  set profileVisibility(value: ProfileVisibility) {
    this.data.profile.profileVisibility = value;
  }

  get startedDiving(): string | undefined {
    return this.data.profile.startedDiving;
  }
  set startedDiving(value: string | undefined) {
    this.data.profile.startedDiving = value;
  }

  async save(): Promise<void> {
    const profile = assertValid<ProfileData>(
      this.data.profile,
      ProfileDataSchema,
    );
    await this.agent.put(`/api/profiles/${this.username}`).send(profile);
    this.data.profile = profile;
  }

  toJSON(): ProfileData {
    return assertValid(this.data.profile, ProfileDataSchema);
  }
}
