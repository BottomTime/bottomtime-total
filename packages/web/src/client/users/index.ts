export * from './interfaces';

export * from './default-user';
export * from './default-user-manager';
