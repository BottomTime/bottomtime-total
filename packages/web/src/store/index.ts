export * from './alerts.store';
export * from './current-user.store';
export * from './operators.store';
export * from './dive-sites.store';
export * from './error.store';
export * from './friends.store';
export * from './log-entries.store';
export * from './profiles.store';
export * from './tanks.store';
export * from './toasts.store';
