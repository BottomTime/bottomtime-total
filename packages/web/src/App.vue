<template>
  <div v-if="isLoading"></div>
  <div v-else>
    <SnackBar />
    <NavBar />
    <RouterView />
    <PageFooter />
    <CookieWarning />
  </div>
</template>

<script lang="ts" setup>
import { onBeforeMount, ref } from 'vue';

import CookieWarning from './components/main/CookieWarning.vue';
import { Dispatch, useStore } from '@/store';
import { inject } from '@/helpers';
import NavBar from '@/components/main/NavBar.vue';
import PageFooter from '@/components/main/PageFooter.vue';
import SnackBar from '@/components/main/SnackBar.vue';
import { ApiClientKey } from '@/injection-keys';

const client = inject(ApiClientKey);
const store = useStore();
const isLoading = ref(true);

onBeforeMount(async () => {
  try {
    const currentUser = await client.users.getCurrentUser();
    await store.dispatch(Dispatch.SignInUser, currentUser);
  } catch (error) {
    console.error(error);
  } finally {
    isLoading.value = false;
  }
});
</script>

<style>
@import './assets/bulmaswatch.min.css';
</style>
