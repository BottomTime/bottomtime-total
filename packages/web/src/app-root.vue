<template>
  <SnackBar />
  <div
    id="modal"
    class="font-content text-grey-900 bg-blue-200 dark:text-grey-100 dark:bg-blue-950"
  ></div>
  <div id="drawer"></div>

  <div
    class="md:container mx-auto font-content text-grey-900 bg-blue-200 dark:text-grey-100 dark:bg-blue-950 rounded-b-xl shadow-md shadow-white dark:shadow-grey-700 opacity-95"
  >
    <!-- Placeholder element. Modal dialogs will be "teleported" here to get the z-indexing right. -->

    <section class="mt-16 p-4">
      <RouterView></RouterView>
    </section>
    <PageFooter />
  </div>
  <NavBar />
</template>

<script setup lang="ts">
import { onBeforeMount } from 'vue';

import { ToastType } from './common';
import NavBar from './components/core/nav-bar.vue';
import PageFooter from './components/core/page-footer.vue';
import SnackBar from './components/core/snack-bar.vue';
import { Config } from './config';
import { useErrors, useToasts } from './store';

const toasts = useToasts();
const errors = useErrors();

onBeforeMount(() => {
  if (!Config.isSSR) {
    if (errors.renderError) {
      if (!Config.isProduction) {
        /* eslint-disable-next-line no-console */
        console.error(errors.renderError);
      }
      toasts.toast({
        id: 'server-prerender-error',
        message:
          'A server error has occurred and the page may not have loaded correctly. We apologize for the inconvenience. Please try again later as we investigate the issue.',
        type: ToastType.Error,
      });
    }
  }
});
</script>
