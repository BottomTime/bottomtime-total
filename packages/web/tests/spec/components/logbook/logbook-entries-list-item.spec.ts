import { ComponentMountingOptions, mount } from '@vue/test-utils';

import { Pinia, createPinia } from 'pinia';
import StarRatingStub from 'tests/stubs/star-rating-stub.vue';

import LogbookEntriesListItem from '../../../../src/components/logbook/logbook-entries-list-item.vue';
import { useCurrentUser } from '../../../../src/store';
import { FullLogEntry, MinimalLogEntry } from '../../../fixtures/log-entries';
import { BasicUser } from '../../../fixtures/users';

const Checkbox = `[data-testid="toggle-${FullLogEntry.id}"]`;
const EditButton = `[data-testid="edit-entry-${FullLogEntry.id}"]`;
const DeleteButton = `[data-testid="delete-entry-${FullLogEntry.id}"]`;
const Title = `[data-testid="select-${FullLogEntry.id}"]`;

describe('LogbookEntriesListItem component', () => {
  let pinia: Pinia;
  let currentUser: ReturnType<typeof useCurrentUser>;
  let opts: ComponentMountingOptions<typeof LogbookEntriesListItem>;

  beforeEach(() => {
    pinia = createPinia();
    currentUser = useCurrentUser(pinia);
    currentUser.user = BasicUser;
    opts = {
      props: {
        editMode: true,
        entry: FullLogEntry,
      },
      global: {
        plugins: [pinia],
        stubs: {
          StarRating: StarRatingStub,
        },
      },
    };
  });

  it('will render correctly with full logbook entry data', () => {
    const wrapper = mount(LogbookEntriesListItem, opts);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('will render correctly for partial logbook entry data', async () => {
    const wrapper = mount(LogbookEntriesListItem, opts);
    await wrapper.setProps({ entry: MinimalLogEntry });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('will show checkbox when in edit mode', () => {
    const wrapper = mount(LogbookEntriesListItem, opts);
    expect(wrapper.find(Checkbox).isVisible()).toBe(true);
  });

  it('will hide checkbox when not in edit mode', async () => {
    const wrapper = mount(LogbookEntriesListItem, opts);
    await wrapper.setProps({ editMode: false });
    expect(wrapper.find(Checkbox).exists()).toBe(false);
  });

  it('will emit "toggle-select" event when checkbox is clicked', async () => {
    const wrapper = mount(LogbookEntriesListItem, opts);
    await wrapper.get(Checkbox).setValue(true);
    expect(wrapper.emitted('toggle-select')).toEqual([[FullLogEntry]]);
  });

  it('will emit "edit" event if edit button is clicked', async () => {
    const wrapper = mount(LogbookEntriesListItem, opts);
    await wrapper.get(EditButton).trigger('click');
    expect(wrapper.emitted('edit')).toEqual([[FullLogEntry]]);
  });

  it('will emit "delete" event if delete button is clicked', async () => {
    const wrapper = mount(LogbookEntriesListItem, opts);
    await wrapper.get(DeleteButton).trigger('click');
    expect(wrapper.emitted('delete')).toEqual([[FullLogEntry]]);
  });

  it('will emit "highlight" event if title text is clicked', async () => {
    const wrapper = mount(LogbookEntriesListItem, opts);
    await wrapper.get(Title).trigger('click');
    expect(wrapper.emitted('highlight')).toEqual([[FullLogEntry]]);
  });
});
