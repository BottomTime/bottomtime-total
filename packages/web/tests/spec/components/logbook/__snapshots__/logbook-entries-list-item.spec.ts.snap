// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LobbookEntriesListItem component will render correctly for partial logbook entry data 1`] = `
"<li class="flex space-x-3 min-h-24 items-center even:bg-blue-300/40 even:dark:bg-blue-900/40 rounded-md p-4">
  <div class="min-w-16 flex">
    <!--v-if-->
    <!--v-if-->
  </div>
  <div class="flex flex-col gap-3 grow">
    <p class="flex flex-col gap-0.5 md:flex-row md:gap-2 items-center md:items-baseline w-full"><button class="text-link hover:text-link-hover text-2xl" data-testid="select-1d3d41c9-794a-4280-8335-65f503a3bee7" type="button">
        <!--v-if-->January 31, 2023 9:16 AM
      </button><span class="italic text-sm">America/Los_Angeles</span></p>
    <div class="flex flex-col justify-start gap-0 lg:flex-row lg:justify-between">
      <div class="flex space-x-2">
        <p class="font-bold min-w-40 text-right lg:min-w-0 lg:text-left"> Duration: </p><span class="italic">44.1min</span>
      </div>
      <!--v-if-->
    </div>
  </div>
</li>"
`;

exports[`LobbookEntriesListItem component will render correctly with full logbook entry data 1`] = `
"<li class="flex space-x-3 min-h-24 items-center even:bg-blue-300/40 even:dark:bg-blue-900/40 rounded-md p-4">
  <div class="min-w-16 flex">
    <!--v-if-->
    <p class="font-bold">#12</p>
  </div>
  <div class="flex flex-col gap-3 grow">
    <p class="flex flex-col gap-0.5 md:flex-row md:gap-2 items-center md:items-baseline w-full"><button class="text-link hover:text-link-hover text-2xl" data-testid="select-4ca3e53c-1364-40b2-a6db-ce8e92380b1f" type="button">
        <!--v-if-->January 31, 2023 9:16 AM
      </button><span class="italic text-sm">Europe/Dublin</span></p>
    <div class="flex flex-col justify-start gap-0 lg:flex-row lg:justify-between">
      <div class="flex space-x-2">
        <p class="font-bold min-w-40 text-right lg:min-w-0 lg:text-left"> Bottom time / Duration: </p><span class="italic">77.8min / 88.3min</span>
      </div>
      <div class="flex space-x-2">
        <p class="font-bold min-w-40 text-right lg:min-w-0 lg:text-left"> Max depth: </p><span class="italic">10.15 m</span>
      </div>
    </div>
  </div>
</li>"
`;
